File: src/core/microsoft.ts
import fetch from 'node-fetch';
import { Translator } from './translator';

export class MicrosoftTranslator implements Translator {
  name!:string;

  targetLang: string;
  apiKey: string;

  constructor(targetLang: string, apiKey: string) {
    this.targetLang = targetLang;
    this.apiKey = apiKey;
  }

  async translate(text: string): Promise<string> {
    let url = new URL('https://api.cognitive.microsofttranslator.com/translate');
    url.searchParams.append('api-version', '3.0');
    url.searchParams.append('to', this.targetLang.split('-')[0]);
    console.log(text);

    let response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify([{
        text,
      }]),
      headers: {
        'Ocp-Apim-Subscription-Key': this.apiKey,
        'Content-type': 'application/json; charset=UTF-8',
      }
    });
    let body = await response.json();
    if (!response.ok) {
      throw new Error(body.error.message);
    }
    return body[0].translations[0].text;
  }
}
MicrosoftTranslator.prototype.name = 'Microsoft';

File: server/src/translator/microsoft.rs
use super::Translator;
use async_trait::async_trait;

pub struct MicrosoftTranslator {
    api_key: String,
    target_lang: String,
}

impl MicrosoftTranslator {
    pub fn new(api_key: String, target_lang: String) -> Self {
        Self {
            api_key,
            target_lang,
        }
    }
}

#[async_trait]
impl Translator for MicrosoftTranslator {
    fn name(&self) -> &'static str {
        "Microsoft"
    }

    async fn translate(&self, text: &str) -> anyhow::Result<String> {
        const API_URL: &str = "https://api.cognitive.microsofttranslator.com/translate";

        #[derive(serde::Serialize)]
        struct Request<'a> {
            text: &'a str,
        }

        #[derive(serde::Deserialize)]
        struct TransResult {
            text: String,
        }

        #[derive(serde::Deserialize)]
        #[allow(dead_code)]
        struct Response {
            translations: Vec<TransResult>,
        }

        #[derive(serde::Deserialize)]
        #[allow(dead_code)]
        struct ApiError {
            code: u32,
            message: String,
        }

        #[derive(serde::Deserialize)]
        #[allow(dead_code)]
        #[serde(untagged)]
        enum ApiResponse {
            Error { error: ApiError },
            Success(Vec<Response>),
        }

        let api_url = reqwest::Url::parse_with_params(
            API_URL,
            &[("api-version", "3.0"), ("to", &*self.target_lang)],
        )
        .unwrap();
        let client = reqwest::Client::new();
        let body: ApiResponse = client
            .post(api_url)
            .header("Ocp-Apim-Subscription-Key", &*self.api_key)
            .json(&vec![Request { text }])
            .send()
            .await?
            .json()
            .await?;

        match body {
            ApiResponse::Success(mut response) => Ok(response
                .pop()
                .ok_or_else(|| anyhow::anyhow!("Unexpected API result"))?
                .translations
                .pop()
                .ok_or_else(|| anyhow::anyhow!("Unexpected API result"))?
                .text),
            ApiResponse::Error { error } => {
                anyhow::bail!("Error {}: {}", error.code, error.message)
            }
        }
    }
}
